src/snowflake/cli/_plugins/nativeapp/sf_sql_facade.py

```diff
@@ -0,0 +1,199 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from contextlib import contextmanager
+
+from cryptography.utils import cached_property
+from snowflake.cli._plugins.nativeapp.sf_facade_exceptions import (
+    CouldNotUseObjectError,
+    UserScriptError,
+    handle_unclassified_error,
+)
+from snowflake.cli.api.constants import ObjectType
+from snowflake.cli.api.entities.common import get_sql_executor
+from snowflake.cli.api.errno import (
+    DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED,
+    NO_WAREHOUSE_SELECTED_IN_SESSION,
+)
+from snowflake.cli.api.project.util import to_identifier
+from snowflake.cli.api.sql_execution import SqlExecutor
+from snowflake.connector import ProgrammingError
+
+
+class SnowflakeSQLFacade:
+    def __init__(self, sql_executor: SqlExecutor | None = None):
+        self._sql_executor = (
+            sql_executor if sql_executor is not None else get_sql_executor()
+        )
+
+    @cached_property
+    def _log(self):
+        return logging.getLogger(__name__)
+
+    def _use_object(self, object_type: ObjectType, name: str):
+        """
+        Call sql to use snowflake object with error handling
+        @param object_type: ObjectType, type of snowflake object to use
+        @param name: object name, has to be a valid snowflake identifier.
+        """
+        try:
+            self._sql_executor.execute_query(f"use {object_type} {name}")
+        except ProgrammingError as err:
+            if err.errno == DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED:
+                raise CouldNotUseObjectError(object_type, name) from err
+            else:
+                handle_unclassified_error(err, f"Failed to use {object_type} {name}.")
```

Comment:
I think this implies that we'll raise an UnknownConnectorError in this case. That feels a little strange. I would have expected some kind of "CommandExecutionFailure", "InvalidSQLError" or something like that, otherwise we're discarding valuable information. It's a ProgrammingError, so we know the SQL was invalid. We just suspect that we're at fault, and not the user.

Reply:
+1, feels like we might be falling into the same trap as the original `generic_sql_error_handler` but in a different form

Reply:
@sfc-gh-bdufour good callout to add handling for ProgrammingErrors specifically. As the handler is checking errors from higher specificity -> more general . ProgrammingError -> DatabaseError -> Error. 

@sfc-gh-mchok I'll have to disagree.  generic_sql_handler does somewhat the opposite of this. It handles known errors, whereas here the goal is to have a final catch-all filter after we've exhausted specific errors in each context the call is made.